server/channels/app/user.go

```diff
@@ -1038,6 +1038,58 @@ func (a *App) userDeactivated(rctx request.CTX, userID string) *model.AppError {
 		rctx.Logger().Warn("unable to remove auth data by user id", mlog.Err(nErr))
 	}
 
+	if err = a.softDeleteUserDMChannels(rctx, userID, model.GetMillis()); err != nil {
+		rctx.Logger().Warn("Failed to archive DM channels for deactivated user", mlog.Err(err))
+	}
+
+	return nil
+}
+
+func (a *App) softDeleteUserDMChannels(rctx request.CTX, userID string, deleteTimestamp int64) *model.AppError {
+	fromID := ""
+	const pageSize = 200
+
+	for {
+		channels, err := a.Srv().Store().Channel().GetChannelsByUser(userID, false /* includeDeleted */, 0, pageSize, fromID)
+		if err != nil {
+			return model.NewAppError("softDeleteUserDMChannels", "app.channel.get_channels_by_user.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
+		}
+		if len(channels) == 0 {
+			break
+		}
+
+		for _, channel := range channels {
+			fromID = channel.Id
+
+			if channel.Type != model.ChannelTypeDirect {
+				continue
+			}
+			if channel.DeleteAt != 0 {
+				continue
+			}
+
+			if err := a.Srv().Store().Channel().SetDeleteAt(channel.Id, deleteTimestamp, deleteTimestamp); err != nil {
```

Comment:
Why not use `App.DeleteChannel` method? 

https://github.com/mattermost/mattermost-server/blob/d78d59babeb994106e305531de50b8d515427396/server/channels/app/channel.go#L1498